gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactorJobPropCreator.java

```diff
@@ -95,59 +109,169 @@ MRCompactorJobPropCreator build() {
   }
 
   protected final String topic;
+  protected final double priority;
   protected final Path topicInputDir;
+  protected final Path topicInputLateDir;
   protected final Path topicOutputDir;
   protected final Path topicOutputLateDir;
-  protected final Path topicTmpDir;
+  protected final Path topicTmpOutputDir;
   protected final FileSystem fs;
-  protected final boolean deduplicate;
   protected final State state;
+  protected final boolean deduplicate;
+
+  // Whether we should recompact the input folders if new data files are found in the input folders.
+  protected final boolean recompactFromInputPaths;
+
+  // Whether we should recompact the output folders if there are late data files in the output '_late' folder.
+  // If this is set to true, input folders of the datasets will be ignored. The output folders and the
+  // output '_late' folders will be used as input to compaction jobs.
+  protected final boolean recompactFromOutputPaths;
 
   protected MRCompactorJobPropCreator(Builder<?> builder) {
     this.topic = builder.topic;
+    this.priority = builder.priority;
     this.topicInputDir = builder.topicInputDir;
+    this.topicInputLateDir = builder.topicInputLateDir;
     this.topicOutputDir = builder.topicOutputDir;
     this.topicOutputLateDir = builder.topicOutputLateDir;
-    this.topicTmpDir = builder.topicTmpDir;
+    this.topicTmpOutputDir = builder.topicTmpOutputDir;
     this.fs = builder.fs;
-    this.deduplicate = builder.deduplicate;
     this.state = builder.state;
+    this.deduplicate =
+        this.state.getPropAsBoolean(MRCompactor.COMPACTION_DEDUPLICATE, MRCompactor.DEFAULT_COMPACTION_DEDUPLICATE);
+    this.recompactFromInputPaths =
+        this.state.getPropAsBoolean(MRCompactor.COMPACTION_RECOMPACT_FROM_INPUT_FOR_LATE_DATA,
+            MRCompactor.DEFAULT_COMPACTION_RECOMPACT_FROM_INPUT_FOR_LATE_DATA);
+    this.recompactFromOutputPaths = this.state.getPropAsBoolean(MRCompactor.COMPACTION_RECOMPACT_FROM_DEST_PATHS,
+        MRCompactor.DEFAULT_COMPACTION_RECOMPACT_FROM_DEST_PATHS);
   }
 
-  protected List<State> createJobProps() throws IOException {
-    List<State> emptyJobProps = Lists.newArrayList();
+  protected List<Dataset> createJobProps() throws IOException {
     if (!this.fs.exists(this.topicInputDir)) {
       LOG.warn("Input folder " + this.topicInputDir + " does not exist. Skipping topic " + this.topic);
-      return emptyJobProps;
+      return ImmutableList.<Dataset> of();
+    }
+
+    Dataset dataset = new Dataset.Builder().withTopic(this.topic).withPriority(this.priority)
+        .withInputPath(this.recompactFromOutputPaths ? this.topicOutputDir : this.topicInputDir)
```

Comment:
With this setting, on-demand recompaction will only work for `topicOutputDir`, but not for `topicInputDir`. What about the late files in `topicInputDir`, i.e., hourly_late? When will they be recompacted?